### DISCUSSION TITLE
Access Credentials in Custom Tools.

### DISCUSSION THREAD

**Discussion by yavisht** [FlowiseAI/Flowise#688] [Q&A]
Are credentials exposed to custom tools? Is there a way to access them instead of hardcoding in the function?

**Comment by HenryHengZJ** [comment]
yes for now. we'll make the improvment for you to access credential within custom tool function

**Reply by kashbadami** [reply]
@HenryHengZJ has this been done already?

**Reply by yavisht** [reply]
@HenryHengZJ  - If we can use Credentials in custom tools, can you please share some steps? It would be nice to get it by using a syntax like this:
```
$credentials.<credential-name>
```
@kashbadami - I am not sure if we can access credentials but you can pull in a value from a custom variable for now.

Further reading
- https://docs.flowiseai.com/using-flowise/variables
- https://docs.flowiseai.com/integrations/langchain/tools/custom-tool#pass-variables-to-function

From what I know so far,

You can use properties specified in Input Schema as variables with prefix $:
```
Property = userid
Variable = $userid
```
You can get default flow config:
```
$flow.sessionId
$flow.chatId
$flow.chatflowId
$flow.input
```
You can get custom variables: 
```$vars.<variable-name>```

**Reply by Jrakru** [reply]
**UPDATE:  Nevermind my answer.  I did not uncomment the E2B.credential file. It works. Maybe you can pass your credential that way.**

Is this why I cannot use the `-Create New-` for E2B tool credential or is it something completely different?

![image](https://github.com/user-attachments/assets/b5124b2e-37c6-4ed8-a2e4-3ee049bfdfd4)

I get the following error message:

`
Object { stack: "Ze@http://localhost:4000/assets/index-B8igdIt_.js:279:30154\nUq@http://localhost:4000/assets/index-B8igdIt_.js:281:1045\nb@http://localhost:4000/assets/index-B8igdIt_.js:281:4237\nEventHandlerNonNull*Qq</<@http://localhost:4000/assets/index-B8igdIt_.js:281:4309\nQq<@http://localhost:4000/assets/index-B8igdIt_.js:281:3214\nT_@http://localhost:4000/assets/index-B8igdIt_.js:283:512\npromise callback*request@http://localhost:4000/assets/index-B8igdIt_.js:283:4036\nEp.prototype[t]@http://localhost:4000/assets/index-B8igdIt_.js:283:4477\nDR/<@http://localhost:4000/assets/index-B8igdIt_.js:279:24888\nAe@http://localhost:4000/assets/CredentialListDialog-D0A1RglP.js:11:1063\nP@http://localhost:4000/assets/CredentialInputHandler-1gRoQTsO.js:1:3244\nonCreateNew@http://localhost:4000/assets/CredentialInputHandler-1gRoQTsO.js:1:4084\nonChange@http://localhost:4000/assets/CredentialInputHandler-1gRoQTsO.js:1:2035\nSe@http://localhost:4000/assets/TextField-D0QM7QQV.js:1:6139\nqe@http://localhost:4000/assets/TextField-D0QM7QQV.js:1:6364\nSt@http://localhost:4000/assets/TextField-D0QM7QQV.js:1:9287\nr3@http://localhost:4000/assets/index-B8igdIt_.js:59:9864\ni3@http://localhost:4000/assets/index-B8igdIt_.js:59:10018\na3@http://localhost:4000/assets/index-B8igdIt_.js:59:10075\noS@http://localhost:4000/assets/index-B8igdIt_.js:59:31462\nZ2@http://localhost:4000/assets/index-B8igdIt_.js:59:31881\nCg/<@http://localhost:4000/assets/index-B8igdIt_.js:59:36794\nby@http://localhost:4000/assets/index-B8igdIt_.js:62:36957\nS2@http://localhost:4000/assets/index-B8igdIt_.js:59:8999\nCg@http://localhost:4000/assets/index-B8igdIt_.js:59:33168\nK0@http://localhost:4000/assets/index-B8igdIt_.js:59:17378\nS3@http://localhost:4000/assets/index-B8igdIt_.js:59:17160\nEventListener.handleEvent*J2@http://localhost:4000/assets/index-B8igdIt_.js:59:32677\n_g@http://localhost:4000/assets/index-B8igdIt_.js:59:32077\ntc/<@http://localhost:4000/assets/index-B8igdIt_.js:59:32243\ntc@http://localhost:4000/assets/index-B8igdIt_.js:59:32188\nO4@http://localhost:4000/assets/index-B8igdIt_.js:62:19760\nQC@http://localhost:4000/assets/index-B8igdIt_.js:62:40098\nqC@http://localhost:4000/assets/index-B8igdIt_.js:62:39851\nL4@http://localhost:4000/assets/index-B8igdIt_.js:62:39720\ncp@http://localhost:4000/assets/index-B8igdIt_.js:62:39570\nNS@http://localhost:4000/assets/index-B8igdIt_.js:62:36685\nxa@http://localhost:4000/assets/index-B8igdIt_.js:60:3279\nF4@http://localhost:4000/assets/index-B8igdIt_.js:62:41286\n$a@http://localhost:4000/assets/index-B8igdIt_.js:62:40278\nGC@http://localhost:4000/assets/index-B8igdIt_.js:62:35735\nP@http://localhost:4000/assets/index-B8igdIt_.js:47:1601\nZ@http://localhost:4000/assets/index-B8igdIt_.js:47:1968\nEventHandlerNonNull*@http://localhost:4000/assets/index-B8igdIt_.js:47:2138\n@http://localhost:4000/assets/index-B8igdIt_.js:47:3914\n", message: "Request failed with status code 500", name: "AxiosError", code: "ERR_BAD_RESPONSE", config: {…}, request: XMLHttpRequest, response: {…} }`
